### PR TITLE
Thermal expansion Eigenstrain with AD

### LINKED ISSUE #13091: AD Eigenstrains
## Reason
Eigenstrains have yet to receive the AD treatment in Tensor Mechanics. They need to be implemented in order to use thermal expansion and other swelling models

## Design
- [ ] ADComputeEigenstrainBase
- [ ] ADComputeThermalExpansionEigenstrainBase
- [ ] ADComputeThermalExpansionEigenstrain

## Impact
Extend the capability of AD in TM

### PR DESCRIPTION
**Pull request by dschwen**
These classes won't have to use the `DerivativeMaterialInterface` (ADMaterials and ADKernels have no optional properties, as all properties are used for the residual calculation!). Instead I use the aptly named `DerivativeMaterialPropertyNameInterface` directly :-D

Closes #13091

### CODE REVIEW

File: framework/include/materials/DerivativeFunctionMaterialBase.h [hunk 1]
```diff
@@ -27,8 +27,8 @@ InputParameters validParams<DerivativeFunctionMaterialBase>();
  *
  * A DerivativeFunctionMaterialBase provides numerous material properties which contain
  * the free energy and its derivatives. The material property names are
```
Comment (tophmatthews) [inline framework/include/materials/DerivativeFunctionMaterialBase.h]: since this is in the framework now, should this say free energy?

File: modules/tensor_mechanics/test/tests/thermal_expansion/tests [hunk 2]
```diff
@@ -35,12 +35,35 @@
...
+    input = 'ad_constant_expansion_stress_free_temp.i'
+    exodiff = 'ad_constant_expansion_stress_free_temp_out.e'
+    requirement = 'The tensor mechanics module shall have the capability to '
+                  'calculate the eigenstrain tensor resulting from isotropic '
+                  'thermal expansion with an initial strain due to the '
+                  'difference between the stress free temperature and initial '
+                  'temperature of the material.'
+    design = 'ADComputeThermalExpansionEigenstrain.md'
+    issues = '#13091'
+  [../]
```
Comment (tophmatthews) [inline modules/tensor_mechanics/test/tests/thermal_expansion/tests]: `PetscJacobianTester`?

**Review by tophmatthews — DISMISSED** [review]
Awesome! You even did the thermal expansion pieces I was thinking I had to do. Thanks @dschwen

File: modules/tensor_mechanics/test/tests/thermal_expansion/tests (hunk 2 above)
Comment (dschwen) [reply modules/tensor_mechanics/test/tests/thermal_expansion/tests]: Do you _still_ doubt the power of the mighty AD? ;-)

File: framework/include/materials/DerivativeFunctionMaterialBase.h (hunk 1 above)
Comment (dschwen) [reply framework/include/materials/DerivativeFunctionMaterialBase.h]: fair point!

File: modules/tensor_mechanics/test/tests/thermal_expansion/tests (hunk 2 above)
Comment (dschwen) [reply modules/tensor_mechanics/test/tests/thermal_expansion/tests]: Hm, glad I added those tests now. The existing tests are messed up (strain=small & ComputeFiniteStrainElasticStress in one and strain=finite & ComputeFiniteStrainElasticStress in the other)

**Review by tophmatthews — DISMISSED** [review]

**Comment by dschwen** [comment]
Renaming the `DerivativeMaterialPropertyNameInterface` methods showed me that they are used more widely than I thought (in some apps). I've added deprecated aliases with the old name back in. The `mooseDeprecated` call will probably still make those apps fail. I guess that means updating apps...

File: modules/tensor_mechanics/doc/content/source/materials/ADComputeThermalExpansionEigenstrain.md [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+
+where $T$ is the current temperature, $T_{stress\_free}$ is the stress free
+temperature, and $\boldsymbol{I}$ is the identity matrix. Jacobian contributions
+are computed using forward mode automatic differentiation.
+
+## Example Input File Syntax
+
+!listing modules/tensor_mechanics/test/tests/thermal_expansion/ad_constant_expansion_stress_free_temp.i block=Materials/thermal_expansion_strain
+
+The `eigenstrain_name` parameter value must also be set for the strain
```
Comment (bwspenc) [inline modules/tensor_mechanics/doc/content/source/materials/ADComputeThermalExpansionEigenstrain.md]: eigenstrain_name -> eigenstrain_names

File: modules/tensor_mechanics/test/tests/thermal_expansion/constant_expansion_coeff.i [hunk 4]
```diff
@@ -36,7 +36,7 @@
   [./TensorMechanics]
     [./Master]
       [./all]
-        strain = SMALL
+        strain = FINITE
```
Comment (bwspenc) [inline modules/tensor_mechanics/test/tests/thermal_expansion/constant_expansion_coeff.i]: Why the change from small? The thermal eigenstrains are all based on linear assumptions, so my guess is that it was made small strain originally because it's easier to directly compare against a known solution.

File: modules/tensor_mechanics/doc/content/source/materials/ADComputeThermalExpansionEigenstrain.md (hunk 3 above)
Comment (bwspenc) [reply modules/tensor_mechanics/doc/content/source/materials/ADComputeThermalExpansionEigenstrain.md]: Fix in the MD file that you ripped this off of too.

**Review by bwspenc — CHANGES_REQUESTED** [review]

**Review by bwspenc — APPROVED** [review]
